### PR TITLE
Add rotary input support for Android platform

### PR DESCRIPTION
**Pull request by AlekseyKapustyanenko**
This  PR adds Rotary input support for wear os devices. It converts RotaryInput to vertical mouse wheel event. Without it Godot does not recognise this event at all.

### CODE REVIEW

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 1]
```diff
@@ -487,10 +487,10 @@ static boolean handleMouseEvent(final MotionEvent event) {
 		float verticalFactor = 0;
 		float horizontalFactor = 0;
 
-		if (event.isFromSource(InputDevice.SOURCE_ROTARY_ENCODER)){
+		if (event.isFromSource(InputDevice.SOURCE_ROTARY_ENCODER)) {
 			// If event came from RotaryEncoder(Bezel or Crown rotate event on wear os smart watches)
```
Comment (akien-mga) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:491]:
```suggestion
			// If event came from RotaryEncoder (Bezel or Crown rotate event on Wear OS smart watches),
```

**Comment by akien-mga** [comment]
Assuming the implementation will be accepted by the @godotengine/android team, one step prior to merging would be to squash the commits. See [PR workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) for instructions.

**Comment by AlekseyKapustyanenko** [comment]
Here is issue in proposal repo https://github.com/godotengine/godot-proposals/issues/9062

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 2]
```diff
@@ -484,8 +484,17 @@ static boolean handleMouseEvent(final MotionEvent event) {
...
 
-		final float verticalFactor = event.getAxisValue(MotionEvent.AXIS_VSCROLL);
-		final float horizontalFactor = event.getAxisValue(MotionEvent.AXIS_HSCROLL);
+		float verticalFactor = 0;
+		float horizontalFactor = 0;
+
+		if (event.isFromSource(InputDevice.SOURCE_ROTARY_ENCODER)) {
+			// If event came from RotaryEncoder (Bezel or Crown rotate event on Wear OS smart watches),
+			// convert it to vertical mouse wheel event.
+			verticalFactor = event.getAxisValue(MotionEvent.AXIS_SCROLL);
```
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:493]: The documentation you linked to has additional logic that is not replicated here; is there a reason to leave that logic out?

You should also provide the ability for users to select whether this should be converted to a vertical event or to a horizontal event using project settings.

**Review by m4gr3d — CHANGES_REQUESTED** [review]

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java (hunk 2 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:493]: Thank you for review. Yes you are right I need to make negation for axis value. All other logic in [this documentation](https://developer.android.com/training/wearables/user-input/rotary-input#java) connected with getting some settings from configuration, like getting scroll velocity and applying RotaryInput to UI control.

> You should also provide the ability for users to select whether this should be converted to a vertical event or to a horizontal event using project settings.

I have concerns about ability to change direction of this input from vertical to horizontal through project settings. As I understand we should convert some specific inputs to another more general inputs. So here I conver RotaryInput to mouse wheel input(most of miсe have only one wheel for vertical scrolling). If I added such settings It would be very disappointed, because in the same game running on PC mouse wheel would cause vertical scrolling but on smart watches crown wheel or bezel wheel would cause horizontal scrolling. Anyway In the same game such event can cause as vertical as horizontal scrolling in different scenes. So how to handle this events and which axis should be chaned by this event should be defined in specific places(inside controls or scene or other object scripts) in my opinion.
So if I have not convinced you, let me know and I will add this setting.

Comment (m4gr3d) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java]: For a cross-platform game/app, I'd agree that consistency is key and for that reason the default value for the project setting should be set to vertical to simulate similar interaction on other platforms. 

However a developer can also decide to target wear os devices only and as such should not be restricted by other platforms conventions. For that reason, we should provide the option to override how the rotary input is interpreted.

Comment (AlekseyKapustyanenko) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java]: Added option to change RotaryInpur axis through project settings.

File: main/main.cpp [hunk 3]
```diff
@@ -2857,6 +2857,7 @@ Error Main::setup2() {
 
 		GLOBAL_DEF_BASIC("input_devices/pointing/android/enable_long_press_as_right_click", false);
 		GLOBAL_DEF_BASIC("input_devices/pointing/android/enable_pan_and_scale_gestures", false);
+		GLOBAL_DEF_RST_NOVAL(PropertyInfo(Variant::STRING, "input_devices/pointing/android/rotary_input_scroll_axis", PROPERTY_HINT_ENUM, "Y,X"), "Y");
```
[the remark this answers is not in the document]
Comment (m4gr3d) [reply main/main.cpp:2860]: I agree the axis value should be an `int` instead of a `string`. This matches the pattern used in the codebase.

As for the labels, I think using `Horizontal, Vertical` would be more clear.

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt [hunk 4]
```diff
@@ -272,6 +273,11 @@ open class GodotEditor : GodotActivity() {
 	 */
 	protected open fun enablePanAndScaleGestures() =
 		java.lang.Boolean.parseBoolean(GodotLib.getEditorSetting("interface/touchscreen/enable_pan_and_scale_gestures"))
+	/**
+	 * Define which axis of mouse wheel event will be used for mapping RotaryInput.
+	 */	
+	protected open fun setRotaryInputAxis() =
```
Comment (m4gr3d) [inline platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt:279]: This should be `getRotaryInputAxis()`.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 5]
```diff
@@ -70,6 +73,8 @@ public class GodotInputHandler implements InputManager.InputDeviceListener {
 	 * Used to decide whether mouse capture can be enabled.
 	 */
 	private int lastSeenToolType = MotionEvent.TOOL_TYPE_UNKNOWN;
+	
+	private static String rotaryInputAxis = AXIS_Y;
```
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:77]: You can remove `static`.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 6]
```diff
@@ -484,8 +496,22 @@ static boolean handleMouseEvent(final MotionEvent event) {
...
-		final float verticalFactor = event.getAxisValue(MotionEvent.AXIS_VSCROLL);
-		final float horizontalFactor = event.getAxisValue(MotionEvent.AXIS_HSCROLL);
+		float verticalFactor = 0;
+        float horizontalFactor = 0;
+        
+        // If event came from RotaryEncoder (Bezel or Crown rotate event on Wear OS smart watches),
+        // convert it to vertical mouse wheel event.
+        if (event.isFromSource(InputDevice.SOURCE_ROTARY_ENCODER)) {        	
+        	if (rotaryInputAxis.equals(AXIS_X)) {
+        		horizontalFactor = event.getAxisValue(MotionEvent.AXIS_SCROLL);
```
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:506]: The negation should be applied here as well.

**Review by m4gr3d — CHANGES_REQUESTED** [review]

File: main/main.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mickeon) [reply main/main.cpp:2860]: `Horizontal,Vertical`. No spacing, it is not trimmed when visualized. Notably, there's really going to ever be just two options so the setting could also be reworked into a toggle.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java (hunk 5 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:77]: No I can not. This property is used in `handleMouseEvent` function which is static.

File: doc/classes/InputEventMouse.xml [hunk 7]
```diff
@@ -5,6 +5,7 @@
 	</brief_description>
 	<description>
 		Stores general information about mouse events.
+		[b]Note:[/b] On Wear os devices RotaryInput maps to MOUSE_BUTTON_UP and MOUS_BUTTON_DOWN events by default. This behavior can be chacnged to MOUSE_BUTTON_UP and MOUS_BUTTON_DOWN events with [member ProjectSettings.input_devices/pointing/android/rotary_input_scroll_axis] project setting.
```
[the remark this answers is not in the document]
Comment (AlekseyKapustyanenko) [reply doc/classes/InputEventMouse.xml]: Fixed. Moved to InputEventMouseButton

File: doc/classes/ProjectSettings.xml [hunk 8]
```diff
@@ -1358,6 +1358,9 @@
 		<member name="input_devices/pointing/android/enable_pan_and_scale_gestures" type="bool" setter="" getter="" default="false">
 			If [code]true[/code], multi-touch pan and scale gestures are enabled on Android devices.
 		</member>
+		<member name="input_devices/pointing/android/rotary_input_scroll_axis" type="String" setter="" getter="" default="Y">
+			Defines which axis of mouse wheel RotaryInput will be mapped on Wear OS devices.
```
[the remark this answers is not in the document]
Comment (AlekseyKapustyanenko) [reply doc/classes/ProjectSettings.xml]: Fixed

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt (hunk 4 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt]: Fixed

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java (hunk 6 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java]: fixed

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 9]
```diff
@@ -101,6 +106,13 @@ public void enableLongPress(boolean enable) {
 	public void enablePanningAndScalingGestures(boolean enable) {
 		this.godotGestureHandler.setPanningAndScalingEnabled(enable);
 	}
+	
+	/**
+	 * Set RotaryInput axis. This is Y by default.
```
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:111]: This javadoc should be updated to match the added documentation.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 10]
```diff
@@ -484,8 +496,22 @@ static boolean handleMouseEvent(final MotionEvent event) {
 		final float tiltX = (float)-Math.sin(orientation) * tiltMult;
 		final float tiltY = (float)Math.cos(orientation) * tiltMult;
 
-		final float verticalFactor = event.getAxisValue(MotionEvent.AXIS_VSCROLL);
-		final float horizontalFactor = event.getAxisValue(MotionEvent.AXIS_HSCROLL);
+		float verticalFactor = 0;
+		float horizontalFactor = 0;
+
+		// If event came from RotaryEncoder (Bezel or Crown rotate event on Wear OS smart watches),
+		// convert it to vertical mouse wheel event.
```
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:503]: The `vertical` in this sentence can be removed.

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt [hunk 11]
```diff
@@ -272,6 +273,11 @@ open class GodotEditor : GodotActivity() {
 	 */
 	protected open fun enablePanAndScaleGestures() =
 		java.lang.Boolean.parseBoolean(GodotLib.getEditorSetting("interface/touchscreen/enable_pan_and_scale_gestures"))
+	/**
+	 * Define which axis of mouse wheel event will be used for mapping RotaryInput.
```
Comment (m4gr3d) [inline platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt:277]: The javadoc should be updated to match the added documentation.

File: doc/classes/ProjectSettings.xml [hunk 12]
```diff
@@ -1358,6 +1358,9 @@
 		<member name="input_devices/pointing/android/enable_pan_and_scale_gestures" type="bool" setter="" getter="" default="false">
 			If [code]true[/code], multi-touch pan and scale gestures are enabled on Android devices.
 		</member>
+		<member name="input_devices/pointing/android/rotary_input_scroll_axis" type="int" setter="" getter="" default="Vertical">
```
Comment (m4gr3d) [inline doc/classes/ProjectSettings.xml:1361]: The value for `default` should be `1`

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt [hunk 13]
```diff
@@ -110,6 +110,7 @@ open class GodotEditor : GodotActivity() {
 		super.onGodotSetupCompleted()
 		val longPressEnabled = enableLongPressGestures()
 		val panScaleEnabled = enablePanAndScaleGestures()
+		val rotaryInputAxis = getRotaryInputAxis()
```
Comment (m4gr3d) [inline platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt:113]: The `rotaryInputAxis` value needs to be passed to the `GodotInputHandler` below on line 121.

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt [hunk 14]
```diff
@@ -272,6 +273,11 @@ open class GodotEditor : GodotActivity() {
 	 */
 	protected open fun enablePanAndScaleGestures() =
 		java.lang.Boolean.parseBoolean(GodotLib.getEditorSetting("interface/touchscreen/enable_pan_and_scale_gestures"))
+	/**
+	 * Define which axis of mouse wheel event will be used for mapping RotaryInput.
+	 */	
+	protected open fun getRotaryInputAxis() =
+		GodotLib.getEditorSetting("input_devices/pointing/android/rotary_input_scroll_axis")
```
Comment (m4gr3d) [inline platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt:280]: `GodotLib.getEditorSetting(...)` returns a `String`, so you need to use `Integer.parseInt(...)` to parse it to an `int` value.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 15]
```diff
@@ -56,6 +56,9 @@
  */
 public class GodotInputHandler implements InputManager.InputDeviceListener {
 	private static final String TAG = GodotInputHandler.class.getSimpleName();
+	
+	private static final int VERTICAL_AXIS = 1;
+	private static final int HORIZONTAL_AXIS = 0;
```
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:61]: Let's update these constants to `ROTARY_INPUT_VERTICAL_AXIS` and `ROTARY_INPUT_HORIZONTAL_AXIS` for clarity.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 16]
```diff
@@ -101,6 +106,13 @@ public void enableLongPress(boolean enable) {
 	public void enablePanningAndScalingGestures(boolean enable) {
 		this.godotGestureHandler.setPanningAndScalingEnabled(enable);
 	}
+	
+	/**
+	 * Set RotaryInput axis. This is Y by default.
+	 */
+	public void setRotaryInputAxis(int axis) {
+    	rotaryInputAxis = axis;
+    }
```
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java:115]: Looks like there is an issue with indentation here.

**Review by m4gr3d — CHANGES_REQUESTED** [review]

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java (hunk 9 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java]: Fixed.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java (hunk 10 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java]: Fixed.

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt (hunk 11 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt]: Fixed.

File: doc/classes/ProjectSettings.xml (hunk 12 above)
Comment (AlekseyKapustyanenko) [reply doc/classes/ProjectSettings.xml]: Fixed.

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt (hunk 13 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt:113]: Fixed.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java (hunk 15 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java]: Fixed.

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java (hunk 16 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java]: Fixed. And All others indentation errors.

File: main/main.cpp [hunk 17]
```diff
@@ -2857,7 +2857,7 @@ Error Main::setup2() {
 
 		GLOBAL_DEF_BASIC("input_devices/pointing/android/enable_long_press_as_right_click", false);
 		GLOBAL_DEF_BASIC("input_devices/pointing/android/enable_pan_and_scale_gestures", false);
-
+		GLOBAL_DEF_BASIC(PropertyInfo(Variant::INT, "input_devices/pointing/android/rotary_input_scroll_axis", PROPERTY_HINT_ENUM, "Horizontal,Vertical"), "1");
```
[the remark this answers is not in the document]
Comment (AlekseyKapustyanenko) [reply main/main.cpp]: Thank you. Fixed

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt (hunk 13 above)
Comment (m4gr3d) [reply platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt:113]: Actually,  I made a mistake; the changes in the `GodotEditor` class are not needed because the editor just used the `Godot` object which applies the setting for it.
So you should be able to revert the changes in this file.

**Review by m4gr3d — CHANGES_REQUESTED** [review]
Looks great! 

There is one set of changes that need to be reverted then it'll be good to merge!

**Review by Mickeon — APPROVED** [review]
<details>
<summary>Documentation is good now.</summary>

</details>

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt [hunk 18]
```diff
@@ -118,6 +118,7 @@ open class GodotEditor : GodotActivity() {
 			godotFragment?.godot?.renderView?.inputHandler?.apply {
 				enableLongPress(longPressEnabled)
 				enablePanningAndScalingGestures(panScaleEnabled)
+				setRotaryInputAxis(rotaryInputAxis)
```
Comment (m4gr3d) [inline platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt:121]: You can also revert this change for the `GodotEditor`.

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt [hunk 19]
```diff
@@ -272,6 +273,11 @@ open class GodotEditor : GodotActivity() {
 	 */
 	protected open fun enablePanAndScaleGestures() =
 		java.lang.Boolean.parseBoolean(GodotLib.getEditorSetting("interface/touchscreen/enable_pan_and_scale_gestures"))
+	/**
+	 * On Wear OS devices, defines which axis of the mouse wheel rotary input is mapped to.
+	 */
+	protected open fun getRotaryInputAxis() =
+		Integer.parseInt(GodotLib.getEditorSetting("input_devices/pointing/android/rotary_input_scroll_axis"))
```
Comment (m4gr3d) [inline platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt:280]: You can also revert this change for the `GodotEditor`.

**Review by m4gr3d — CHANGES_REQUESTED** [review]

File: platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt (hunk 18 above)
Comment (AlekseyKapustyanenko) [reply platform/android/java/editor/src/main/java/org/godotengine/editor/GodotEditor.kt]: fixed

File: doc/classes/ProjectSettings.xml [hunk 20]
```diff
@@ -1358,6 +1358,9 @@
 		<member name="input_devices/pointing/android/enable_pan_and_scale_gestures" type="bool" setter="" getter="" default="false">
 			If [code]true[/code], multi-touch pan and scale gestures are enabled on Android devices.
 		</member>
+		<member name="input_devices/pointing/android/rotary_input_scroll_axis" type="int" setter="" getter="" default="1">
+			On Wear OS devices, defines which axis of the mouse wheel rotary input is mapped to.
```
Comment (Calinou) [inline doc/classes/ProjectSettings.xml:1362]:
```suggestion
			On Wear OS devices, defines which axis of the mouse wheel rotary input is mapped to. This rotary input is performed by rotating the physical or virtual bezel on a smartwatch.
```

**Review by m4gr3d — APPROVED** [review]
Looks great!

File: platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java [hunk 21]
```diff
@@ -102,6 +107,13 @@ public void enablePanningAndScalingGestures(boolean enable) {
 		this.godotGestureHandler.setPanningAndScalingEnabled(enable);
 	}
 
+	/**
+	 * On Wear OS devices, sets which axis of the mouse wheel rotary input is mapped to. This is 1(vertical axis) by default.
```
[the remark this answers is not in the document]
Comment (AlekseyKapustyanenko) [reply platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java]: Fixed

File: doc/classes/ProjectSettings.xml (hunk 20 above)
Comment (AlekseyKapustyanenko) [reply doc/classes/ProjectSettings.xml]: fixed

**Review by Calinou — APPROVED** [review]
Code and docs look good to me. I don't have a setup to test this locally though.[^1]

[^1]: I acquired a Galaxy Watch 6 the other day, but didn't look at sideloading apps on it yet.

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by AlekseyKapustyanenko** [comment]
> Code and docs look good to me. I don't have a setup to test this locally though.[1](#user-content-fn-1-2d97edef1035191e0ed53080270ff69e)
> 
> ## Footnotes
> 1. I acquired a Galaxy Watch 6 the other day, but didn't look at sideloading apps on it yet. [↩](#user-content-fnref-1-2d97edef1035191e0ed53080270ff69e)

I have checked this solution on my smartwatches. So let me know if I could help you to check it on your side.
